stmhal: Pick AF for I2C pins dynamically

dhylands opened the pull request:
This also adds I2C2 and I2C3 definitions for the Nucleo 401 board
and fixes up pins.csv

On PYBV10 this increases the flash size by 16 bytes. I suspect that having more peripherals use this mechanism will reduce the code size.

I tested this on a PYBOARD (I2C1 & 2) my CERB40 (I2C1,2,3), and on a NUCLEO-F401RE (I2C1,2,3) board.

dpgeorge commented:
Thanks, but please make separate commits for the changes to the board files and the I2C stuff.

dhylands committed:
stmhal: Add a function for setting the pin alternate function

mp_hal_gpio_set_af will search for a given function and unit
and set the alternate function to the alternate function index
found.

dhylands committed:
stmhal: NUCLEO 401 - Add a bunch of missing pins and reorder the pins

This groups the pins for a port together and puts them in numerical
order. It also adds ARDUINO pin names.

dhylands committed:
stmhal: CERB40 - Add pins defines for I2C3

dhylands committed:
stmhal: NUCLEO 401 - Add definitions for I2C2 and I2C3

dhylands commented:
rebased and split into several commits
